File: leetcode/algorithms/tree/binary-tree-preorder-traversal.java
/*
    leetcode
    category: "Algorithms"
    problem title: "Binary Tree Preorder Traversal"
*/


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    private void populatePreorderList(TreeNode node,List<Integer> list){
        if(node==null){
            return;
        }
        
        list.add(node.val);
        populatePreorderList(node.left,list);
        populatePreorderList(node.right,list);
        
    }
    
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> preorderList  = new ArrayList<Integer>();
        populatePreorderList(root,preorderList);
        
        return preorderList;
        
    }
}

File: leetcode/algorithms/tree/binary-tree-preorder-traversal.ts
/*
    leetcode
    category: "Algorithms"
    problem title: "Binary Tree Preorder Traversal"
*/

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function preorder(root: TreeNode | null, order: number[]): void {
    if(root===null){
        return;
    }
    order.push(root.val);;
    preorder(root.left,order);
    preorder(root.right,order);
}

function preorderTraversal(root: TreeNode | null): number[] {
    const order: number[] = [];
    preorder(root,order);
    
    return order;

};